kythe/extractors/openjdk11/java_wrapper/java_wrapper.go

```diff
@@ -0,0 +1,215 @@
+	}
+	return "", os.ErrNotExist
+}
+
+func findExtractorJar() (string, error) {
+	if val := os.Getenv(extractorJarVar); len(val) > 0 {
+		return val, checkPath(val)
+	}
+	return bazel.Runfile(extractorJarPath)
+}
+
+func extractorArgs(args []string, jar string) []string {
+	isJavac := false
+	var result []string
+	for len(args) > 0 {
+		var a string
+		var v string
+		switch a, args = shift(args); a {
+		case "-m", "--module":
+			v, args = shift(args)
+			if !strings.HasSuffix(v, ".javac.Main") {
+				isJavac = false
+				break
+			}
+			isJavac = true
+			result = append(result,
+				"--add-exports=jdk.compiler.interim/com.sun.tools.javac.main=ALL-UNNAMED",
+				"--add-exports=jdk.compiler.interim/com.sun.tools.javac.util=ALL-UNNAMED",
+				"--add-exports=jdk.compiler.interim/com.sun.tools.javac.file=ALL-UNNAMED",
+				"--add-exports=jdk.compiler.interim/com.sun.tools.javac.api=ALL-UNNAMED",
+				"--add-exports=jdk.compiler.interim/com.sun.tools.javac.code=ALL-UNNAMED",
+				"-jar", jar, "-Xprefer:source")
+		case "--add-modules", "--limit-modules":
+			v, args = shift(args)
+			result = append(result, a, v+",java.logging,java.sql")
+		case "--doclint-format":
+			_, args = shift(args)
+		case "-Werror":
+		default:
+			switch {
+			case strings.HasPrefix(a, "-Xplugin:depend"), strings.HasPrefix(a, "-Xlint:"), strings.HasPrefix(a, "-Xdoclint"):
+			case strings.HasPrefix(a, "-Xmx"):
+				result = append(result, "-Xmx3G")
+			default:
+				result = append(result, a)
+			}
+		}
+	}
+	// As we can only do anything meaningful with Java compilations,
+	// but wrap the java binary, don't attempt to exact other invocations.
+	if !isJavac {
+		return nil
+	}
+	return result
+}
+
+func extractorEnv() []string {
+	env := os.Environ()
+	env = setEnvDefault(env, kytheRootVar, ".")
+	env = setEnvDefault(env, kytheOutputVar, outputDir())
+	env = setEnvDefault(env, kytheTargetVar, moduleName())
+	env = setEnvDefault(env, kytheCorpusVar, "openjdk11")
+	return env
+}
+
+func setEnvDefault(env []string, key, def string) []string {
+	if val := os.Getenv(key); len(val) == 0 {
+		env = append(env, key+"="+def)
+	}
+	return env
+}
+
+func shift(args []string) (string, []string) {
+	if len(args) > 0 {
+		return args[0], args[1:]
+	}
+	return "", nil
+}
+
+func init() {
+	setupRunfiles()
+}
+
+func main() {
+	java, err := findJavaCommand()
+	if err != nil {
+		log.Fatalf("ERROR: unable to find java executable: %v", err)
+	}
+	jar, err := findExtractorJar()
+	if err != nil {
+		log.Fatalf("ERROR: unable to find extractor jar: %v", err)
+	}
+	if args := extractorArgs(os.Args[1:], jar); len(args) > 0 {
+		cmd := exec.Command(java, args...)
+		cmd.Env = extractorEnv()
+		log.Printf("*** Extracting: %s", moduleName())
+		if output, err := cmd.CombinedOutput(); err != nil {
+			w, err := os.Create(filepath.Join(outputDir(), moduleName()+".err"))
+			if err != nil {
+				log.Printf("Error creating error log for module %s: %v", moduleName(), err)
+			}
+			defer w.Close()
```

Comment:
Ordinarily you shouldn't close a writable file without checking the result. In this case it might be OK since it's just collecting human-readable diagnostics anyway.

However: The defer is probably not what you want in this case, since this function will not return. Probably the descriptor will get closed on exec anyway, but for intent I suggest putting this at the end of the `if` block without a defer.

Reply:
Done.